components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+        PreUpdateProfileAction preUpdateProfileAction = (PreUpdateProfileAction) actionExecutionContext.getAction();
+
+        return new ActionExecutionRequest.Builder()
+                .actionType(getSupportedActionType())
+                .event(getEvent(userActionContext, preUpdateProfileAction))
+                .build();
+    }
+
+    private Event getEvent(UserActionContext userActionContext, PreUpdateProfileAction preUpdateProfileAction)
+            throws ActionExecutionRequestBuilderException {
+
+        PreUpdateProfileEvent.Builder eventBuilder = new PreUpdateProfileEvent.Builder();
+        eventBuilder.initiatorType(getInitiatorType());
+        eventBuilder.action(PreUpdateProfileEvent.Action.UPDATE);
+        eventBuilder.request(getPreUpdateProfileRequest(userActionContext));
+        eventBuilder.tenant(getTenant());
+
+        UniqueIDUserStoreManager userStoreManager = getUserStoreManager();
+        eventBuilder.user(getUser(userActionContext, preUpdateProfileAction, userStoreManager));
+        eventBuilder.userStore(getUserStore(userActionContext, userStoreManager));
+
+        return eventBuilder.build();
+    }
+
+    private PreUpdateProfileEvent.FlowInitiatorType getInitiatorType() throws ActionExecutionRequestBuilderException {
+
+        Flow flow = Optional.ofNullable(IdentityContext.getThreadLocalIdentityContext().getFlow())
+                .orElseThrow(() -> new ActionExecutionRequestBuilderException("Unknown flow."));
+
+        switch (flow.getInitiatingPersona()) {
+            case ADMIN:
+                return PreUpdateProfileEvent.FlowInitiatorType.ADMIN;
+            case APPLICATION:
+                return PreUpdateProfileEvent.FlowInitiatorType.APPLICATION;
+            case USER:
+                return PreUpdateProfileEvent.FlowInitiatorType.USER;
+            default:
+                throw new ActionExecutionRequestBuilderException(
+                        "Unknown initiator type for the flow: " + flow.getName());
+        }
+    }
+
+    private PreUpdateProfileRequest getPreUpdateProfileRequest(UserActionContext userActionContext)
+            throws ActionExecutionRequestBuilderException {
+
+        UserActionRequestDTO userActionRequestDTO = userActionContext.getUserActionRequestDTO();
+
+        if (userActionRequestDTO.getClaims() == null || userActionRequestDTO.getClaims().isEmpty()) {
+            return null;
+        }
+
+        PreUpdateProfileRequest.Builder preUpdateProfileRequestBuilder = new PreUpdateProfileRequest.Builder();
+
+        for (Map.Entry<String, Object> claimEntry : userActionRequestDTO.getClaims().entrySet()) {
+            if (isRoleOrGroupClaim(claimEntry.getKey())) {
+                 /* Skip role and group claims.
+                 Profile update extension should not share role and group claims as claims updating.
+                 Updating roles or groups should be a separate workflow.*/
+                continue;
+            }
```

Comment:
The logic is correct, but do we get roles and groups updates for profile update event?. Just tried to see whether this logic is required or not if no claim updates are recieving.

Reply:
When it comes to PUT requests this layer received role claims from SCIM layer.
For the same reason you mentioned that at a profile update roles and groups should not be updated or should be communicated as being updated to the extension this logic is added.
I think it's safer to have this than not having this

Reply:
`I think it's safer to have this than not having this` Yes agree on that.